bi-kafka/kafka.go

```
@@ -0,0 +1,214 @@
+func (p *Producer) Produce(ctx context.Context, msg *Message, version string) error {
+	if msg == nil {
+		return fmt.Errorf("message cannot be nil")
+	}
+
+	if msg.Topic == "" {
+		return fmt.Errorf("topic cannot be empty")
+	}
+	msg.Headers = append(msg.Headers, sarama.RecordHeader{
+		Key:   []byte("version"),
+		Value: []byte(version),
+	})
+	// Create Sarama message
+	saramaMsg := &sarama.ProducerMessage{
+		Topic:   msg.Topic,
+		Key:     sarama.ByteEncoder(msg.Key),
+		Value:   sarama.ByteEncoder(msg.Value),
+		Headers: msg.Headers,
+	}
+
+	if !msg.Timestamp.IsZero() {
+		saramaMsg.Timestamp = msg.Timestamp
+	}
+
+	// Send message with context
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case p.producer.Input() <- saramaMsg:
+		// Message sent to async producer
+		return nil
+	}
+}
+
+// ProduceBatch sends multiple messages to Kafka
+func (p *Producer) ProduceBatch(ctx context.Context, messages []*Message, version string) error {
+	if len(messages) == 0 {
+		return nil
+	}
+
+	saramaMessages := make([]*sarama.ProducerMessage, len(messages))
+	for i, msg := range messages {
+		if msg == nil {
+			return fmt.Errorf("message at index %d is nil", i)
+		}
+
+		if msg.Topic == "" {
+			return fmt.Errorf("topic cannot be empty for message at index %d", i)
+		}
+		msg.Headers = append(msg.Headers, sarama.RecordHeader{
+			Key:   []byte("version"),
+			Value: []byte(version),
+		})
+		saramaMessages[i] = &sarama.ProducerMessage{
+			Topic:   msg.Topic,
+			Key:     sarama.ByteEncoder(msg.Key),
+			Value:   sarama.ByteEncoder(msg.Value),
+			Headers: msg.Headers,
+		}
+
+		if !msg.Timestamp.IsZero() {
+			saramaMessages[i].Timestamp = msg.Timestamp
+		}
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		// Send all messages to async producer
+		for _, saramaMsg := range saramaMessages {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case p.producer.Input() <- saramaMsg:
+				// Message sent
+			}
```

we should fail fast, if buffer is full